File: frontend/src/types/notification.ts
/**
 * Types for in-app notifications
 */
import dayjs from 'dayjs';

export type NotificationTypeT =
  | 'user_published_schedule'
  | 'user_created_request'
  | 'user_accepted_request'
  | 'user_denied_request'
  | 'user_created_assignment'
  | 'user_updated_assignment'
  | 'user_deleted_assignment'
  | 'user_received_team_invite'
  | 'user_accepted_team_invite'
  | 'user_removed_from_team'
  | 'user_left_team'
  | 'user_created_direct_swap'
  | 'user_accepted_direct_swap'
  | 'user_refused_direct_swap'
  | 'user_created_open_swap'
  | 'user_bid_open_swap'
  | 'user_selected_bid_open_swap'
  | 'user_selected_other_bid_open_swap'
  | 'swap_ready_for_review'
  | 'user_validated_swap'
  | 'user_denied_swap'
  | 'user_reversed_swap'
  | 'campaign_request_deadline_set'
  | 'campaign_request_deadline_reminder'
  | 'campaign_request_deadline_extended';

export type NotificationT = {
  id: string;
  userId: string;
  teamId: string;
  type: NotificationTypeT;
  eventData: Record<string, string>;
  read: boolean;
  createdAt: dayjs.Dayjs;
  updatedAt: dayjs.Dayjs;
  readAt: dayjs.Dayjs | null;
  seenAt: dayjs.Dayjs | null;
};

export type NotificationKey =
  | 'user_published_schedule'
  | 'user_created_request'
  | 'user_accepted_request'
  | 'user_denied_request'
  | 'user_created_assignment'
  | 'user_updated_assignment'
  | 'user_deleted_assignment'
  | 'user_received_team_invite'
  | 'user_accepted_team_invite'
  | 'user_removed_from_team'
  | 'user_left_team'
  | 'user_created_direct_swap'
  | 'user_accepted_direct_swap'
  | 'user_refused_direct_swap'
  | 'user_created_open_swap'
  | 'user_bid_open_swap'
  | 'user_selected_bid_open_swap'
  | 'user_selected_other_bid_open_swap'
  | 'swap_ready_for_review'
  | 'user_validated_swap'
  | 'user_denied_swap'
  | 'user_reversed_swap'
  | 'campaign_request_deadline_set'
  | 'campaign_request_deadline_reminder'
  | 'campaign_request_deadline_extended';

export const NOTIFICATION_KEYS: NotificationKey[] = [
  'user_published_schedule',
  'user_created_request',
  'user_accepted_request',
  'user_denied_request',
  'user_created_assignment',
  'user_updated_assignment',
  'user_deleted_assignment',
  'user_received_team_invite',
  'user_accepted_team_invite',
  'user_removed_from_team',
  'user_left_team',
  'user_created_direct_swap',
  'user_accepted_direct_swap',
  'user_refused_direct_swap',
  'user_created_open_swap',
  'user_bid_open_swap',
  'user_selected_bid_open_swap',
  'user_selected_other_bid_open_swap',
  'swap_ready_for_review',
  'user_validated_swap',
  'user_denied_swap',
  'user_reversed_swap',
  'campaign_request_deadline_set',
  'campaign_request_deadline_reminder',
  'campaign_request_deadline_extended',
];

export type NotificationCategory = 'schedule' | 'requests' | 'assignments' | 'team' | 'swaps';

export type NotificationKeyMeta = {
  category: NotificationCategory;
};

export const NOTIFICATION_REGISTRY: Record<NotificationKey, NotificationKeyMeta> = {
  user_published_schedule: { category: 'schedule' },
  user_created_request: { category: 'requests' },
  user_accepted_request: { category: 'requests' },
  user_denied_request: { category: 'requests' },
  user_created_assignment: { category: 'assignments' },
  user_updated_assignment: { category: 'assignments' },
  user_deleted_assignment: { category: 'assignments' },
  user_received_team_invite: { category: 'team' },
  user_accepted_team_invite: { category: 'team' },
  user_removed_from_team: { category: 'team' },
  user_left_team: { category: 'team' },
  user_created_direct_swap: { category: 'swaps' },
  user_accepted_direct_swap: { category: 'swaps' },
  user_refused_direct_swap: { category: 'swaps' },
  user_created_open_swap: { category: 'swaps' },
  user_bid_open_swap: { category: 'swaps' },
  user_selected_bid_open_swap: { category: 'swaps' },
  user_selected_other_bid_open_swap: { category: 'swaps' },
  swap_ready_for_review: { category: 'swaps' },
  user_validated_swap: { category: 'swaps' },
  user_denied_swap: { category: 'swaps' },
  user_reversed_swap: { category: 'swaps' },
  campaign_request_deadline_set: { category: 'schedule' },
  campaign_request_deadline_reminder: { category: 'schedule' },
  campaign_request_deadline_extended: { category: 'schedule' },
};

export const NOTIFICATION_CATEGORY_ORDER: NotificationCategory[] = [
  'schedule',
  'requests',
  'assignments',
  'team',
  'swaps',
];

export type ChannelPreferences = { email: boolean; inApp: boolean };

export type NotificationPreferencesT = {
  userId: string;
  preferences: Record<NotificationKey, ChannelPreferences>;
};

export function toNotificationT(data: any): NotificationT {
  return {
    id: data.id,
    userId: data.userId,
    teamId: data.teamId,
    type: data.type as NotificationTypeT,
    eventData: data.eventData ?? {},
    read: data.read ?? false,
    createdAt: dayjs(data.createdAt),
    updatedAt: dayjs(data.updatedAt),
    readAt: data.readAt ? dayjs(data.readAt) : null,
    seenAt: data.seenAt ? dayjs(data.seenAt) : null,
  };
}

export function toNotificationPreferencesT(data: any): NotificationPreferencesT {
  const rawPrefs: Record<string, any> = data.preferences ?? {};
  const preferences = Object.fromEntries(
    NOTIFICATION_KEYS.map((key) => {
      const ch = rawPrefs[key];
      return [key, { email: ch?.email ?? true, inApp: ch?.inApp ?? true }];
    }),
  ) as Record<NotificationKey, ChannelPreferences>;
  return { userId: data.userId, preferences };
}

export function fromNotificationPreferencesT(prefs: NotificationPreferencesT): any {
  return {
    userId: prefs.userId,
    preferences: Object.fromEntries(
      NOTIFICATION_KEYS.map((key) => {
        const ch = prefs.preferences[key];
        return [key, { email: ch.email, inApp: ch.inApp }];
      }),
    ),
  };
}

File: backend/shared/src/shared/schemas/core/notification.py
"""Notification schemas for NSP Pro."""

from dataclasses import dataclass
from datetime import datetime, timezone
from enum import Enum
from typing import Any, Dict, Optional, cast

import humps
from pydantic import BaseModel, Field


class NotificationType(str, Enum):
    """Type of in-app notification."""

    USER_PUBLISHED_SCHEDULE = "user_published_schedule"
    USER_CREATED_REQUEST = "user_created_request"
    USER_ACCEPTED_REQUEST = "user_accepted_request"
    USER_DENIED_REQUEST = "user_denied_request"
    USER_CREATED_ASSIGNMENT = "user_created_assignment"
    USER_UPDATED_ASSIGNMENT = "user_updated_assignment"
    USER_DELETED_ASSIGNMENT = "user_deleted_assignment"
    USER_RECEIVED_TEAM_INVITE = "user_received_team_invite"
    USER_ACCEPTED_TEAM_INVITE = "user_accepted_team_invite"
    USER_REMOVED_FROM_TEAM = "user_removed_from_team"
    USER_LEFT_TEAM = "user_left_team"
    # Swap notification types
    USER_CREATED_DIRECT_SWAP = "user_created_direct_swap"
    USER_ACCEPTED_DIRECT_SWAP = "user_accepted_direct_swap"
    USER_REFUSED_DIRECT_SWAP = "user_refused_direct_swap"
    USER_CREATED_OPEN_SWAP = "user_created_open_swap"
    USER_BID_OPEN_SWAP = "user_bid_open_swap"
    USER_SELECTED_BID_OPEN_SWAP = "user_selected_bid_open_swap"
    USER_SELECTED_OTHER_BID_OPEN_SWAP = "user_selected_other_bid_open_swap"
    SWAP_READY_FOR_REVIEW = "swap_ready_for_review"
    USER_VALIDATED_SWAP = "user_validated_swap"
    USER_DENIED_SWAP = "user_denied_swap"
    USER_REVERSED_SWAP = "user_reversed_swap"
    CAMPAIGN_REQUEST_DEADLINE_SET = "campaign_request_deadline_set"
    CAMPAIGN_REQUEST_DEADLINE_REMINDER = "campaign_request_deadline_reminder"
    CAMPAIGN_REQUEST_DEADLINE_EXTENDED = "campaign_request_deadline_extended"


@dataclass
class Notification:
    """Core notification domain object."""

    id: str
    user_id: str
    team_id: str
    type: NotificationType
    event_data: Dict[str, Any]
    read: bool
    created_at: datetime
    updated_at: datetime
    read_at: Optional[datetime] = None
    seen_at: Optional[datetime] = None

    def to_dto(self) -> "NotificationDTO":
        data = {
            "id": self.id,
            "userId": self.user_id,
            "teamId": self.team_id,
            "type": self.type.value,
            "eventData": self.event_data,
            "read": self.read,
            "createdAt": self.created_at.isoformat(),
            "updatedAt": self.updated_at.isoformat(),
            "readAt": self.read_at.isoformat() if self.read_at else None,
            "seenAt": self.seen_at.isoformat() if self.seen_at else None,
        }
        return NotificationDTO(**cast(Dict[str, Any], humps.camelize(data)))

    @classmethod
    def from_dto(cls, dto: "NotificationDTO") -> "Notification":
        return cls(
            id=dto.id or "",
            user_id=dto.user_id,
            team_id=dto.team_id,
            type=NotificationType(dto.type),
            event_data=dto.event_data,
            read=dto.read,
            created_at=dto.created_at,
            updated_at=dto.updated_at,
            read_at=dto.read_at,
            seen_at=dto.seen_at,
        )


@dataclass
class NotificationEvent:
    """Carries all data needed to create notifications for one or more users."""

    notification_type: NotificationType
    user_ids: list[str]
    team_id: str
    event_data: Dict[str, Any]


class NotificationDTO(BaseModel):
    """Pydantic DTO for Notification (camelCase, JSON-serialisable)."""

    id: Optional[str] = Field(default=None, alias="id")
    user_id: str = Field(..., alias="userId")
    team_id: str = Field(..., alias="teamId")
    type: str = Field(..., alias="type")
    event_data: Dict[str, Any] = Field(..., alias="eventData")
    read: bool = Field(default=False, alias="read")
    created_at: datetime = Field(
        default_factory=lambda: datetime.now(timezone.utc), alias="createdAt"
    )
    updated_at: datetime = Field(
        default_factory=lambda: datetime.now(timezone.utc), alias="updatedAt"
    )
    read_at: Optional[datetime] = Field(default=None, alias="readAt")
    seen_at: Optional[datetime] = Field(default=None, alias="seenAt")

    model_config = {"populate_by_name": True}
